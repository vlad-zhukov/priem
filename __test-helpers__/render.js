/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import {render} from 'react-testing-library';

export default (element, options) => {
    const ref = React.createRef();
    const el = React.cloneElement(element, {ref});
    const result = render(el, options);
    result.instance = ref.current;
    return result;
};
