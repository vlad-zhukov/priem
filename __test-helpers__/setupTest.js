/* eslint-disable import/no-extraneous-dependencies */

import 'raf/polyfill';
