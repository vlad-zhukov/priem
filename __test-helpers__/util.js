/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import delay from 'delay';
import Priem from '../src/Priem';
import {Container, populateStore} from '../src/Container';

export function testComponent({initialStore, options} = {}) {
    if (initialStore) {
        populateStore(initialStore);
    }

    const ctr = new Container({
        mapPropsToArgs: () => ['foo'],
        promise: value => delay(100, {value}),
        ...options,
    });

    return <Priem sources={{ctr}}>{p => <div>{p.ctr}</div>}</Priem>;
}

export function testComponentNested({initialStore, ctr1Props, ctr2Props} = {}) {
    if (initialStore) {
        populateStore(initialStore);
    }

    const ctr1 = new Container({
        mapPropsToArgs: () => ['foo'],
        promise: value => delay(100, {value}),
        ...ctr1Props,
    });

    const ctr2 = new Container({
        mapPropsToArgs: p => (!p.ctr1 ? null : [p.ctr1, 'bar']),
        promise: (ctr1Value, value) => delay(100, {value: ctr1Value + value}),
        ...ctr2Props,
    });

    return (
        <Priem sources={{ctr1, ctr2}}>
            {props => (
                <Priem sources={{ctr2}} ctr1={props.ctr1}>
                    {p => <div>{p.ctr2}</div>}
                </Priem>
            )}
        </Priem>
    );
}

/* eslint-disable react/no-unused-state */
export class ErrorBoundary extends React.Component {
    constructor() {
        super();
        this.state = {initTime: Date.now(), hasError: null};
    }

    componentDidCatch(error) {
        this.setState({hasError: error, catchTime: Date.now()});
    }

    render() {
        if (this.state.hasError) {
            return null;
        }
        return this.props.children;
    }
}
/* eslint-enable react/no-unused-state */
