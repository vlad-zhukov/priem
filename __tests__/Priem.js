import React from 'react';
import delay from 'delay';
import {mount} from 'enzyme';
import {createSerializer} from 'enzyme-to-json';
import {TestComponentSimple, TestComponentNested, initialStoreTestComponentSimple} from '../__test-helpers__/util';

expect.addSnapshotSerializer(createSerializer({mode: 'deep'}));

function getStateFromSources(sources) {
    return Object.keys(sources).reduce((result, key) => {
        const source = sources[key];
        result[key] = {state: source.state, meta: source.meta};
        return result;
    }, {});
}

it('should render a simple component', async () => {
    const wrapper = mount(<TestComponentSimple />);
    await delay(150);
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
});

it('should rehydrate ssr data', async () => {
    const wrapper = mount(<TestComponentSimple initialState={initialStoreTestComponentSimple} />);
    await delay(150);
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
});

it('should render a nested component', async () => {
    const wrapper = mount(<TestComponentNested />);
    await delay(300);
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(getStateFromSources(wrapper.children().instance()._sources)).toMatchSnapshot();

    wrapper.find('button').simulate('click');
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
    expect(getStateFromSources(wrapper.children().instance()._sources)).toMatchSnapshot();
});
