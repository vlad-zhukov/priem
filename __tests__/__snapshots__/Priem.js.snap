// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should rehydrate ssr data 1`] = `
<div>
  baz
</div>
`;

exports[`should render a nested component 1`] = `
<div>
  2-foobar
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`should render a nested component 2`] = `
Object {
  "container1": Object {
    "meta": Object {
      "args": Array [
        2,
        "foo",
      ],
      "ssr": false,
    },
    "state": Object {
      "fulfilled": true,
      "pending": false,
      "reason": null,
      "refreshing": false,
      "rejected": false,
      "value": "2-foo",
    },
  },
  "syncContainer": Object {
    "meta": undefined,
    "state": Object {
      "counter": 2,
    },
  },
}
`;

exports[`should render a nested component 3`] = `
<div>
  2-foobar
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`should render a nested component 4`] = `
Object {
  "container1": Object {
    "meta": Object {
      "args": Array [
        2,
        "foo",
      ],
      "ssr": false,
    },
    "state": Object {
      "fulfilled": true,
      "pending": false,
      "reason": null,
      "refreshing": true,
      "rejected": false,
      "value": "2-foo",
    },
  },
  "syncContainer": Object {
    "meta": undefined,
    "state": Object {
      "counter": 3,
    },
  },
}
`;

exports[`should render a simple component 1`] = `
<div>
  foo
</div>
`;
